src/auth/tenant.ts

```diff
@@ -0,0 +1,199 @@
+   * @return {?string} The tenant ID corresponding to the resource, null otherwise.
+   */
+  public static getTenantIdFromResourceName(resourceName: string): string | null {
+    // name is of form projects/project1/tenants/tenant1
+    const matchTenantRes = resourceName.match(/\/tenants\/(.*)$/);
+    if (!matchTenantRes || matchTenantRes.length < 2) {
+      return null;
+    }
+    return matchTenantRes[1];
+  }
+
+  /**
+   * Validates a tenant options object. Throws an error on failure.
+   *
+   * @param {any} request The tenant options object to validate.
+   * @param {boolean} createRequest Whether this is a create request.
+   */
+  public static validate(request: any, createRequest: boolean) {
+    const validKeys = {
+      displayName: true,
+      type: true,
+      emailSignInConfig: true,
+    };
+    const label = createRequest ? 'CreateTenantRequest' : 'UpdateTenantRequest';
+    if (!validator.isNonNullObject(request)) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        `"${label}" must be a valid non-null object.`,
+      );
+    }
+    // Check for unsupported top level attributes.
+    for (const key in request) {
+      if (!(key in validKeys)) {
+        throw new FirebaseAuthError(
+          AuthClientErrorCode.INVALID_ARGUMENT,
+          `"${key}" is not a valid ${label} parameter.`,
+        );
+      }
+    }
+    // Validate displayName type if provided.
+    if (typeof request.displayName !== 'undefined' &&
+        !validator.isNonEmptyString(request.displayName)) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        `"${label}.displayName" must be a valid non-empty string.`,
+      );
+    }
+    // Validate type type if provided.
+    if (typeof request.type !== 'undefined' && !createRequest) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        '"Tenant.type" is an immutable property.',
+      );
+    }
+    if (createRequest &&
+        request.type !== 'full_service' &&
+        request.type !== 'lightweight') {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        `"${label}.type" must be either "full_service" or "lightweight".`,
+      );
+    }
+    // Validate emailSignInConfig type if provided.
+    if (typeof request.emailSignInConfig !== 'undefined') {
+      EmailSignInConfig.validate(request.emailSignInConfig);
+    }
+  }
+
+  /**
+   * The Tenant object constructor.
+   *
+   * @param {any} response The server side response used to initialize the Tenant object.
+   * @constructor
+   */
+  constructor(response: any) {
+    const tenantId = Tenant.getTenantIdFromResourceName(response.name);
+    if (!tenantId) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INTERNAL_ERROR,
+        'INTERNAL ASSERT FAILED: Invalid tenant response',
+      );
+    }
+    utils.addReadonlyGetter(this, 'tenantId', tenantId);
```

Comment:
Directly assign to the readonly attributed you have defined above.

Reply:
done